csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,233 @@
+      auto* uop = def->as<UnaryOp>();
+      // TODO: exception to break propagation.
+      if (candidate_check(uop->in())) {
+        stack.emplace(uop, 0);
+        replay_sequence.push_back(uop);
+        continue;
+      }
+      // This will require us having `replayExprWithNewInput` to support binary
+      // ops. } else if (def->isA<BinaryOp>()) {
+      //   auto* bop = def->as<BinaryOp>();
+      //   // TODO: exception to break propagation.
+      //   // TODO: check for broadcast stuff.
+      //   if (candidate_check(bop->lhs()) && candidate_check(bop->rhs())) {
+      //     stack.emplace(uop, 0);
+      //     continue;
+      //   }
+
+      // TODO: adding pad_op
+      // } else if (def->isA<PadOp>()) {
+      //   if (candidate_check(def->input(0))) {
+      //     // NOTE: stopping propagation, we'll merge it with its consumer
+      //     padOp frontier.emplace_back(def, 0); continue;
+      //   }
+    }
+
+    if (edge.val() != pad_op->in()) {
+      // propagation stopped, push entry to frontier
+      frontier.push_back(edge);
+    }
+  }
+
+  if (frontier.empty()) {
+    return nullptr;
+  }
+
+  std::unordered_map<Val*, Val*> replacement_map;
+  // modify pad_op on frontier
+  for (const Edge& edge : frontier) {
+    // insert pad_op
+    // Note: operation with multiple operand would require us to support partial
+    // update in each iteration.
+
+    // const auto width_size = pad_op->inputs().size() - 2;
+    // const auto num_padded_dims = width_size / 2;
+    // std::vector<Val*> pad_width;
+    // pad_width.reserve(width_size);
+    // for (auto i : c10::irange(num_padded_dims)) {
+    //   pad_width.push_back(pad_op->input((num_padded_dims - i)*2));
+    //   pad_width.push_back(pad_op->input((num_padded_dims - i)*2 + 1));
+    // }
+    // cannot use `pad` op, because it would give us symolic iter domain
+    // replacement_map[edge.val()] = pad(edge.val()->as<TensorView>(),
+    // pad_width, pad_op->value());
+
+    NVF_ERROR(
+        edge.val()->getDataType().has_value(), "pad source dtype is missing");
+    // NOTE: the old pad_op is going away from DCE, so it's ok to reuse its
+    // domains
+    auto new_out = IrBuilder::create<TensorView>(
+        IrBuilder::create<TensorDomain>(
+            pad_op->out()->as<TensorView>()->domain()),
+        edge.val()->getDataType().value());
+    IrBuilder::create<PadOp>(
+        new_out,
+        edge.val()->as<TensorView>(),
+        pad_op->getPadWidths(),
+        pad_op->value());
+
+    replacement_map[edge.val()] = new_out;
+
+    // TODO: modify existing pad_op, when its only consumer is a pad_op
+  }
+
+  // propagate to update TensorProxy
+  // need to follow the reverse order from earlier stack traversal.
+  std::reverse(replay_sequence.begin(), replay_sequence.end());
+  for (Expr* e : replay_sequence) {
+    // TODO extend this for multiple inputs.
+    Expr* padded_e = replayExprWithNewInput(e, replacement_map.at(e->input(0)));
```

Comment:
note for myself. I don't think we need this here but rather a newValLike would suffice.